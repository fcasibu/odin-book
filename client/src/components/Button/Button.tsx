import * as React from "react";

const padding = {
    none: "p-0",
    sm: "px-3 py-1",
    md: "px-4 py-2",
    lg: "px-8 py-4",
};

const colors = {
    default: "bg-sky-500",
    success: "bg-green-500",
    danger: "bg-red-500",
    none: "bg-none",
};

const textColors = {
    default: "text-sky-700",
    white: "text-white",
    danger: "text-red-700",
    success: "text-green-700",
};

type Size = keyof typeof padding;
type Variant = keyof typeof colors;
type TextColor = keyof typeof textColors;

type Props = {
    children: React.ReactNode;
    size?: Size;
    variant?: Variant;
    text?: TextColor;
    onClick?: (event: React.MouseEvent) => void;
    type?: "button" | "submit" | "reset";
};

const Button = ({ children, size = "md", variant = "default", text = "white", type = "button", onClick }: Props) => {
    return (
        <button
            className={`${colors[variant]} ${textColors[text]} font-bold rounded-md ${padding[size]} hover:scale-[.98] active:scale-[.95] transition-all ease-linear`}
            onClick={onClick}
            type={type}
        >
            {" "}
            {children}
        </button>
    );
};

export default Button;
