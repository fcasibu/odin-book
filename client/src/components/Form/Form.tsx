import * as React from "react";

type Props = {
    children: React.ReactNode;
    onSubmit?: (event: React.FormEvent) => void;
};

const Form = ({ children, onSubmit }: Props) => {
    return (
        <form onSubmit={onSubmit} className="flex flex-col gap-2 w-full">
            {children}
        </form>
    );
};

export default Form;
