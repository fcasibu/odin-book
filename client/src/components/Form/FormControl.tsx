import * as React from "react";

type Props = {
    children: React.ReactNode;
    label?: string;
    id: string;
    error?: { message: string };
};

export const FormControl = ({ children, label, error, id }: Props) => {
    return (
        <label htmlFor={id}>
            {label && <p className="text-xs text-slate-700 pl-1 pb-1">{label}</p>}
            {children}
            {error && (
                <div role="alert" className="text-xs text-red-500 pl-1 pt-1">
                    {error.message}
                </div>
            )}
        </label>
    );
};
