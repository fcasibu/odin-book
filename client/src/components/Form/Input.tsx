import * as React from "react";
import { FormControl } from "./FormControl";

type Props = {
    type?: "text" | "email" | "password";
    id: string;
    name: string;
    label?: string;
    error?: { message: string };
    value?: string;
    onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
    required?: boolean;
};

export const Input = (props: Props) => {
    const { type, id, name, error, label, value, onChange, required = false } = props;
    return (
        <FormControl id={id} error={error} label={label}>
            <input
                type={type}
                id={id}
                name={name}
                value={value}
                onChange={onChange}
                required={required}
                className="w-full rounded-sm p-2 border-2 focus:outline-sky-500"
            />
        </FormControl>
    );
};
