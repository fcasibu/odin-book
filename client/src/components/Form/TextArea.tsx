import { FormControl } from "./FormControl";

type Props = {
    id: string;
    name: string;
    label?: string;
    error?: { message: string };
    value?: string;
    onChange?: (event: React.ChangeEvent<HTMLTextAreaElement>) => void;
};

export const TextArea = (props: Props) => {
    const { id, name, label, error, value, onChange } = props;
    return (
        <FormControl label={label} id={id} error={error}>
            <textarea
                id={id}
                name={name}
                value={value}
                onChange={onChange}
                className="w-full rounded-sm p-2 border-2 focus:outline-sky-500 resize-none no-scrollbar"
            ></textarea>
        </FormControl>
    );
};
