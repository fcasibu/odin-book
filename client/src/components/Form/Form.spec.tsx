import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Form from "./Form";
import { Input } from "./Input";
import Button from "../Button";
import { TextArea } from "./TextArea";

// TODO: useForm
describe("Form", () => {
    it("should render the form and match the snapshot", () => {
        const { container } = render(
            <Form>
                <Input id="name" name="name" label="Name" />
                <Button type="submit">Submit</Button>
            </Form>
        );
        expect(container).toMatchSnapshot();
    });

    it("should render error message correctly", () => {
        const mockErrors = [{ message: "Invalid Name!" }, { message: "Invalid Post!" }];
        render(
            <>
                <Input id="name" name="name" error={mockErrors[0]} />
                <TextArea id="post" name="post" error={mockErrors[1]} />
            </>
        );

        const errorElements = screen.getAllByRole("alert");
        expect(errorElements[0]).toHaveTextContent(mockErrors[0].message);
        expect(errorElements[1]).toHaveTextContent(mockErrors[1].message);
    });
});
