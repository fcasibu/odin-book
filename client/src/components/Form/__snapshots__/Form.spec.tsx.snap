// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render the form and match the snapshot 1`] = `
<div>
  <form
    class="flex flex-col gap-2 w-full"
  >
    <label
      for="name"
    >
      <p
        class="text-xs text-slate-700 pl-1 pb-1"
      >
        Name
      </p>
      <input
        class="w-full rounded-sm p-2 border-2 focus:outline-sky-500"
        id="name"
        name="name"
        value=""
      />
    </label>
    <button
      class="bg-sky-500 text-white font-bold rounded-md px-4 py-2 hover:opacity-80 active:opacity-100 transition-all ease-linear self-start w-full"
      type="submit"
    >
       
      Submit
    </button>
  </form>
</div>
`;
